sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/FnApiStateAccessor.java

```diff
@@ -600,8 +603,73 @@ public <KeyT, ValueT> MultimapState<KeyT, ValueT> bindMultimap(
   @Override
   public <T> OrderedListState<T> bindOrderedList(
       String id, StateSpec<OrderedListState<T>> spec, Coder<T> elemCoder) {
-    throw new UnsupportedOperationException(
-        "TODO: Add support for a sorted-list state to the Fn API.");
+    return (OrderedListState<T>)
+        stateKeyObjectCache.computeIfAbsent(
+            createOrderedListUserStateKey(id),
+            new Function<StateKey, Object>() {
+              @Override
+              public Object apply(StateKey key) {
+                return new OrderedListState<T>() {
+                  private final OrderedListUserState<T> impl =
+                      createOrderedListUserState(key, elemCoder);
+
+                  @Override
+                  public void clear() {
+                    impl.clear();
+                  }
+
+                  @Override
+                  public void add(TimestampedValue<T> value) {
```

Comment:
Do we have anything that prevents users from using `Long.MAX_VALUE` as the timestamp here? Since all of our ranges are exclusive on the upper bound, I don't think we can read or clear these elements after they've been set (short of clearing the entire list).

Reply:
That's a good point. I think the same problem occurred in 
https://github.com/apache/beam/blob/fd5b1de4f1daaa1c3fb930593f9b27062412d38b/runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryStateInternals.java#L653

Let me open a bug to fix both implementations after this PR.

Reply:
Opened #31593